styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTransactionTest.java

```diff
@@ -94,185 +114,230 @@ public static void tearDownClass() throws Exception {
+    var workflow = TestData.WORKFLOW_WITH_RESOURCES;
+
+    // Store workflow
+    storage.runInTransactionWithRetries(tx -> {
+      tx.store(workflow);
+      return null;
+    });
+
+    // Start a losing transaction that reads, waits for barrier and then stores the workflow
+    var runs = new AtomicInteger();
+    var waiting = new LinkedBlockingQueue<Boolean>();
+    var proceed = new LinkedBlockingQueue<Boolean>();
+    var future = executor.submit(() -> storage.runInTransactionWithRetries(tx -> {
+      runs.incrementAndGet();
+      var wf = tx.workflow(workflow.id());
+      waiting.put(true);
+      proceed.take();
+      tx.store(wf.orElseThrow());
+      return null;
+    }));
+
+    // Execute winning read-store transactions to make each retry of the above transaction fail
+    for (int i = 0; i < RETRY_SETTINGS.getMaxAttempts(); i++) {
+      waiting.take();
+      storage.runInTransactionWithRetries(tx -> {
+        var wf = tx.workflow(workflow.id());
+        tx.store(wf.orElseThrow());
+        return null;
+      });
+      proceed.put(true);
     }
-  }
 
-  @Test
-  public void insertActiveStateShouldFailIfAlreadyExists() throws Exception {
-    DatastoreStorageTransaction tx = new DatastoreStorageTransaction(datastore.newTransaction());
-    tx.writeActiveState(WORKFLOW_INSTANCE1, RUN_STATE);
-    tx.commit();
-    tx = new DatastoreStorageTransaction(datastore.newTransaction());
-    tx.writeActiveState(WORKFLOW_INSTANCE1, RUN_STATE);
+    // Wait for first transaction to give up
     try {
-      tx.commit();
-      fail("Expected exception!");
-    } catch (TransactionException e) {
-      assertThat(e.isAlreadyExists(), is(true));
+      future.get(300, SECONDS);
+      fail("Expected transaction to fail");
+    } catch (ExecutionException e) {
+      var cause = e.getCause();
+      assertThat(cause, instanceOf(DatastoreIOException.class));
+      var dex = ((DatastoreIOException) cause).getCause();
+      assertThat(dex.getCode(), is(10));
     }
+
+    // Verify that it retried as many times as expected before giving up
+    assertThat(runs.get(), is(RETRY_SETTINGS.getMaxAttempts()));
   }
 
+
   @Test
-  public void updateActiveStateShouldFailIfNotFound() throws Exception {
-    DatastoreStorageTransaction tx = new DatastoreStorageTransaction(datastore.newTransaction());
-    tx.updateActiveState(WORKFLOW_INSTANCE1, RUN_STATE);
-    try {
-      tx.commit();
-      fail("Expected exception!");
-    } catch (TransactionException e) {
-      assertThat(e.isNotFound(), is(true));
-    }
+  public void sleepShouldBeInterruptible() throws Exception {
+    var running = new CompletableFuture<String>();
+    var future = executor.submit(() -> {
+      running.complete(null);
+      DatastoreStorage.sleepMillis(Long.MAX_VALUE);
+      return "foobar";
+    });
+    running.join();
+    executor.shutdownNow();
+    var result = future.get(30, SECONDS);
+    assertThat(result, is("foobar"));
```

Comment:
Interesting, I would have expected it to fail with an Interrupted exception

Reply:
The interrupted exception is handled by the `sleepMillis` method: https://github.com/spotify/styx/pull/740/files/d929a7b0e4bee61159efdb673079dededc3cde6c#diff-489bc1d178eed213e99db37d2423a206R993